_database. VO. Configurations. Id for entity

BBSRKhelp opened the pull request:
_Connect the database. Create configurations for all entities. Create an entity responsible for the Id. Add a VO

Commit:
database. VO. Configurations. Id for entity

Connect the database. Create configurations for all entities. Create an entity responsible for the Id. Add a VO

n-cherashev commented on backend/src/PetFamily.API/Program.cs:
```diff
@@ -1,16 +1,44 @@
 
+app.UseHttpsRedirection();
+
+var summaries = new[]
```
Это удалить

n-cherashev commented on backend/src/PetFamily.API/Program.cs:
```diff
@@ -1,16 +1,44 @@
+    "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
+};
+
+app.MapGet("/weatherforecast", () =>
```
И это тоже удалить

n-cherashev commented on backend/src/PetFamily.API/Program.cs:
```diff
@@ -1,16 +1,44 @@
-app.MapGet("/home", () => "Hello, World!");
 app.Run();
+
+record WeatherForecast(DateOnly Date, int TemperatureC, string? Summary)
```
Удалить

n-cherashev commented on backend/src/PetFamily.Domain/SpeciesAggregate/Entities/Breed.cs:
```diff
@@ -0,0 +1,30 @@
+        BreedPet = breedPet;
+    }
+
+    public string BreedPet { get; }
```
Тут лучше назвать подругому, так как класс называется и так Breed, так что свойство не должно повторяться с названием класса, например можно Name, тогда мы свойство будем обращаться к свойству вот так: Breed.Name

n-cherashev commented on backend/src/PetFamily.Domain/SpeciesAggregate/Species.cs:
```diff
@@ -1,23 +1,27 @@
-    public Guid Id { get; private set; }
     public string NameOfSpecies { get; private set; }
-    public IReadOnlyList<Breed> Breeds { get; } = null!;
+    public IReadOnlyList<Breed> Breeds { get; }
```
Тут нужно немного исправть, мы должны добавить приватный List, чтобы мы могли добавлять породы, так как в IReadOnlyList мы неможем добавить ничего.

Должно быть вот так:
private readonly List<Breed> _breeds = [];
public IReadOnlyList<Breed> Breeds => _breeds;

n-cherashev commented on backend/src/PetFamily.Domain/SpeciesAggregate/Species.cs:
```diff
@@ -1,23 +1,27 @@
     }
     
-    public Guid Id { get; private set; }
     public string NameOfSpecies { get; private set; }
```
Тут ненужно лишнее уточнение OfSpecies, и так понятно что это название Species, так как мы находимся в классе Species

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/AppearanceDetails.cs:
```diff
@@ -0,0 +1,64 @@
+
+    public SpeciesId SpeciesId { get; }
+    public BreedId BreedId { get; }
+    public Colour Coloration { get; }
```
в этом ValueObject должны быть только SpeciesId и BreedId

BBSRKhelp replied:
Этот VO создать надо в SpeciesAggregate или в VolunteerAggregate?

n-cherashev replied:
В VolunteerAggregate, так как это относится к Pet

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/AppearanceDetails.cs:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public SpeciesId SpeciesId { get; }
+    public BreedId BreedId { get; }
```
BreedId должен быть типа Guid, так как мы неможем напрямую обратиться к породам другого аггрегата

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+        if (description.Length > 2000)
+            return Result.Failure<Pet>($"'{nameof(Description)}' must be less than 2000 characters.");
+        
+        if (appearanceDetails.SpeciesId == SpeciesId.Empty())
```
Не нужно проводить валидацию ValueObject, так как при создании ValueObject, мы и так его валидируем

n-cherashev replied:
Не убрал это

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+        if (appearanceDetails.SpeciesId == SpeciesId.Empty())
+            return Result.Failure<Pet>($"'{nameof(SpeciesId)}' cannot be null or empty.");
+        
+        if (appearanceDetails.BreedId == BreedId.Empty())
```
убрать

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+        if (appearanceDetails.BreedId == BreedId.Empty())
+            return Result.Failure<Pet>($"'{nameof(BreedId)}' cannot be null or empty.");
+
+        if (appearanceDetails.Weight <= 0 || appearanceDetails.Weight > Constants.MAX_MEDIUM_LOW_TEXT_LENGTH)
```
убрать

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+        if (appearanceDetails.Weight <= 0 || appearanceDetails.Weight > Constants.MAX_MEDIUM_LOW_TEXT_LENGTH)
+            return Result.Failure<Pet>("Weight must be between 0 and 1000");
+        
+        if (appearanceDetails.Height <= 0 || appearanceDetails.Height > Constants.MAX_MEDIUM_LOW_TEXT_LENGTH)
```
убрать

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+        if (appearanceDetails.Height <= 0 || appearanceDetails.Height > Constants.MAX_MEDIUM_LOW_TEXT_LENGTH)
+            return Result.Failure<Pet>("Height must be between 0 and 1000");
+        
+        if (IsNullOrEmpty(healthDetails.HealthInformation))
```
убрать

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+    public AppearanceDetails AppearanceDetails { get; private set; }
+    public HealthDetails HealthDetails { get; private set; }
+    public Address Address { get; private set; }
+    public string PhoneNumber { get; private set; }
```
Лучше добавить отдельный ValueObject для PhoneNumber, так как такое свойство достаточно сложно валидировать (проверка на правильность ввода телефона)

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/Entities/Pet.cs:
```diff
@@ -0,0 +1,102 @@
+        if (IsNullOrEmpty(healthDetails.HealthInformation))
+            return Result.Failure<Pet>("Health information cannot be null or empty.");
+        
+        if (!phoneNumber.Contains(@"^((\(\d{3}\) ?)|(\d{3}-))?\d{3}-\d{4}$"))
```
Это лучше вынести в отдельный ValueObject для номера телефона

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/PetDetails.cs:
```diff
@@ -0,0 +1,30 @@
+        Requisites = requisite;
+    }
+    
+    public List<PetPhoto>? PetPhotos { get; }
```
Тут лучше сделать IReadOnlyList

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/PetDetails.cs:
```diff
@@ -0,0 +1,30 @@
+    }
+    
+    public List<PetPhoto>? PetPhotos { get; }
+    public List<Requisite> Requisites { get; }
```
Тут тоже

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/PetDetails.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+    }
+    
+    private PetDetails(List<PetPhoto> petPhoto, List<Requisite> requisite)
```
Тут лучше сделать IEnumerable вместо List

BBSRKhelp replied:
Чтобы ограничить функции добавления и удаления данных в коллекции? А если так, то полезно ли нам это? Вдруг добавиться еще одна фотография пета? Мы будем полностью сносить список и новый добавлять?

n-cherashev replied:
Я написал именно про параметры в конструкторе.
И да, нам нужно ограничить возможность какого-либо изменения коллекций, так как VO мы неможем менять, мы должны полностью заменять их на новый экземпляр

Так что у самих полей нужно использовать IReadOnlyList

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/PetDetails.cs:
```diff
@@ -0,0 +1,30 @@
+    public List<PetPhoto>? PetPhotos { get; }
+    public List<Requisite> Requisites { get; }
+
+    public static Result<PetDetails> Create(List<PetPhoto> petPhoto, List<Requisite> requisite)
```
Тут не нужен фабричный метод, так как валидировать нечего

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/PetPhoto.cs:
```diff
@@ -0,0 +1,30 @@
+        IsMainPhoto = isMainPhoto;
+    }
+    
+    public string? Path { get; }
```
Лучше убрать '?'

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/VolunteerDetails.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+    }
+    
+    private VolunteerDetails(List<SocialNetwork> socialNetworks, List<Requisite> requisites)
```
Тут сделать также как я написал в PetDetails

n-cherashev commented on backend/src/PetFamily.Infrastructure/ApplicationDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+        optionsBuilder.UseNpgsql(configuration.GetConnectionString(DATABASE));
+        optionsBuilder.UseSnakeCaseNamingConvention();
+        optionsBuilder.UseLoggerFactory(CreateLoggerFactory());
+    }
```
Тут можно добавить ещё:

optionsBuilder.EnableSensitiveDataLogging();

BBSRKhelp committed:
refactoring

Change the name of the column in the species and breeds table. Add a logging method

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/VolunteerDetails.cs:
```diff
@@ -0,0 +1,20 @@
+        Requisites = requisites;
+    }
+    
+    public IEnumerable<SocialNetwork> SocialNetworks { get; }
```
У этих свойств должен быть тип IReadOnlyList.

IEnumerable должен быть у параметров в конструкторе.

n-cherashev commented on backend/src/PetFamily.Domain/VolunteerAggregate/ValueObjects/PetDetails.cs:
```diff
@@ -0,0 +1,22 @@
+        Requisites = requisite;
+    }
+    
+    public IEnumerable<PetPhoto>? PetPhotos { get; }
```
У этих свойств должен быть тип IReadOnlyList.

IEnumerable должен быть у параметров в конструкторе.

Ответил на комментарий твой, там написал зачем это нужно.

n-cherashev reviewed (changes requested):
В сущности Pet, в фабричном методе не исправил + я написал ещё коммент

BBSRKhelp committed:
Alter configuration db

Add a VO for the phone number. Change the constructor of *Details.cs. Delete unnecessary code
